browser-extension/mv2/src/background/background.js

```diff
@@ -974,6 +974,61 @@ BG.Methods.sendMessage = function (messageObject, callback) {
   );
 };
 
+BG.Methods.sendMessageToApp = (messageObject, tabId, timeout = 2000) => {
+  const sendMessage = (messageObject, tabId) => {
+    return new Promise((resolve, reject) => {
+      chrome.tabs.sendMessage(tabId, messageObject, (response) => {
+        if (response) {
+          resolve(response);
+        } else {
+          reject(new Error("No response from app"));
+        }
+      });
+    });
+  };
+
+  return Promise.race([
+    sendMessage(messageObject, tabId),
+    new Promise((_, reject) =>
+      setTimeout(() => reject(new Error("timeout")), timeout)
+    ),
+  ]);
+};
+
+BG.Methods.sendExtensionEvents = (tabId = null, retryCounter = 0) => {
+  let appTabId = tabId;
+  if (!appTabId) {
+    chrome.tabs.query({ url: RQ.configs.WEB_URL + "/*" }, (tabs) => {
+      if (tabs.length > 0) {
+        appTabId = tabs[0].id;
+      } else {
+        BG.isAppOnline = false;
+      }
+    });
+  }
+
+  if (!BG.isAppOnline) {
+    return;
+  }
+
+  const extensionEventsMessage = {
+    action: RQ.EXTENSION_MESSAGES.EXTENSION_EVENTS,
+    events: {}, // it should fetched from a separate function only after all the checks and shouldn't be passed as params
+  };
```

Comment:
We will not get an action from anywhere, will be triggered by a setInterval running in the background script itself

reference code around this commit -- https://github.com/requestly/requestly/pull/454/commits/651785841f82085b8a39d9be0392a583d63fb927